[spark-24442][SQL] Added parameters to control the default number of …

AndrewKL opened the pull request:
…records shown and the truncate length when a user calls .show() on a dataframe

## What changes were proposed in this pull request?

https://issues.apache.org/jira/browse/SPARK-24442

## How was this patch tested?

Unit tests plus local testing.  This change is designed to not modify default behavior unless a user set the following configs...

"spark.sql.show.defaultNumRows" ->default 20 (as it was before)

"spark.sql.show.truncateMaxCharsPerColumn" -> default 30 (as it was before)

Commit:
[spark-24442][SQL] Added parameters to control the default number of records shown and the truncate length when a user calls .show() on a dataframe

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented:
ok to test

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -815,6 +815,24 @@ class Dataset[T] private[sql](
+   * @since 2.3.0
+   */
+  private def numberOfRowsToShow(): Int = {
+    this.sparkSession.conf.get("spark.sql.show.defaultNumRows", "20").toInt
```
I would add it to `SQLConf`

kiszk replied:
+1

maropu replied:
+1

maropu replied:
How about `spark.sql.defaultNumRowsInShow`?

AndrewKL replied:
will do

SparkQA commented:
**[Test build #95254 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/95254/testReport)** for PR 22162 at commit [`ed5991e`](https://github.com/apache/spark/commit/ed5991e53fe404b972c5a0a79febdd0398182cf6).

SparkQA commented:
**[Test build #95254 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/95254/testReport)** for PR 22162 at commit [`ed5991e`](https://github.com/apache/spark/commit/ed5991e53fe404b972c5a0a79febdd0398182cf6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/95254/
Test PASSed.

kiszk commented on sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:
```diff
@@ -969,6 +969,22 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
   }
 
+
+  test("SPARK-2444git stat2 Show should follow spark.show.default.number.of.rows") {
```
nit: Do we remove `git stat2`?

AndrewKL replied:
removed

kiszk commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -815,6 +815,24 @@ class Dataset[T] private[sql](
+   * @since 2.3.0
+   */
+  private def maxCharactersPerColumnToShow(): Int = {
+    this.sparkSession.conf.get("spark.sql.show.truncateMaxCharsPerColumn", "20").toInt
```
ditto

kiszk commented on sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:
```diff
@@ -969,6 +969,22 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
+
+  test("SPARK-2444git stat2 Show should follow spark.show.default.number.of.rows") {
+    withSQLConf("spark.sql.show.defaultNumRows" -> "100") {
+      val ds = (1 to 1000).toDS().as[Int].show
```
Do we need to compare the result with a expected string using `assert`?

maropu replied:
I think its ok to check the output number of rows in show.

AndrewKL replied:
The way show is currently implemented makes it difficult to fully test this.  The truncate and max length parameters are passed into the showString function that would normally be used to test the out put.  unfortunately the parameters and final string is then passed into the println function without an easy way to capture the results.  This could be tested with a spying library but there isn't one in in spark (yet).

kiszk commented on sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:
```diff
@@ -969,6 +969,22 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
+      val ds = (1 to 1000)
+        .map(x => "123456789_123456789_123456789_")
+        .toDS().as[String]
+        .show
```
ditto

HyukjinKwon commented:
ping @AndrewKL

HyukjinKwon commented:
@viirya, @kiszk, @mgaido91 and @maropu, would you be interested in taking this over if this gets inactive for few more days?

mgaido91 commented:
sure @HyukjinKwon, thanks for pinging me anyway.

mgaido91 commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -815,6 +815,24 @@ class Dataset[T] private[sql](
+   * a user specified max number of rows.
+   * @since 2.3.0
+   */
+  private def numberOfRowsToShow(): Int = {
```
I'd remove the `()`

rxin replied:
we shouldn't be adding methods here

mgaido91 commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -815,6 +815,24 @@ class Dataset[T] private[sql](
+   * the show function is called with truncate.
+   * @since 2.3.0
+   */
+  private def maxCharactersPerColumnToShow(): Int = {
```
ditto

mgaido91 commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -815,6 +815,24 @@ class Dataset[T] private[sql](
+  /**
+   * Returns the default max characters per column to show before truncation when
+   * the show function is called with truncate.
+   * @since 2.3.0
```
ditto

AndrewKL replied:
I'll update this.

mgaido91 commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -815,6 +815,24 @@ class Dataset[T] private[sql](
+  /**
+   * Returns the default number of rows to show when the show function is called without
+   * a user specified max number of rows.
+   * @since 2.3.0
```
not needed as this is private, moreover, it'd be since 2.4.0

AndrewKL replied:
I'll update this.

mgaido91 reviewed (commented):
@AndrewKL may you please also change the title making spark uppercase? Thanks.

kiszk commented:
Would someone please take it?
I have less bandwidth next two days since I will be in a training session at my office.

mgaido91 commented:
sure, no worries @kiszk, I can take it if needed. Thanks.

maropu commented:
I have much bandwidh to take it, too. Is it ok to take it over? @mgaido91 not working on this now?

viirya commented:
We should wait @AndrewKL for few days?

maropu commented:
ya, sure.

AndrewKL commented:
Hey Friends,

Sorry for the delayed response.  I've been in the desert 5 hours into the middle of no where.  I'll incorporate  the feed back and update this.

SparkQA commented:
**[Test build #95775 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/95775/testReport)** for PR 22162 at commit [`f93e58e`](https://github.com/apache/spark/commit/f93e58e7471382576840ffc151d57c51b921cb47).

AndrewKL commented:
Hey Friends,  I have updated the commit and fixed most of the issues mentioned.  I do have one remaining issue left which is the testing of the new show functionality.  Currently the provided private function to get the show string takes in the new parameters given by the config so its impossible to test the impact of the config changes.  There's two possible ways to tackle this.  1) leave the tests as is. 2) add in a spy library that would allow me to inspect the input to the showString function.  This would require adding a library to test this.  I wanted to get feed back on the preferred route before I did this.

AndrewKL committed:
Merge branch 'master' into spark-24442

SparkQA commented:
**[Test build #95804 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/95804/testReport)** for PR 22162 at commit [`d85d892`](https://github.com/apache/spark/commit/d85d89200792671b74b0eb71738a5eae58956737).

SparkQA commented:
**[Test build #95804 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/95804/testReport)** for PR 22162 at commit [`d85d892`](https://github.com/apache/spark/commit/d85d89200792671b74b0eb71738a5eae58956737).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/95804/
Test FAILed.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1950,6 +1962,11 @@ class SQLConf extends Serializable with Logging {
   def parallelFileListingInStatsComputation: Boolean =
     getConf(SQLConf.PARALLEL_FILE_LISTING_IN_STATS_COMPUTATION)
 
+  def sqlShowDefaultMaxRecordsToShow: Int = getConf(SQLConf.SQL_SHOW_DEFAULT_MAX_ROWS)
```
`sqlShowDefaultMaxRows`?

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1950,6 +1962,11 @@ class SQLConf extends Serializable with Logging {
 
+  def sqlShowDefaultMaxRecordsToShow: Int = getConf(SQLConf.SQL_SHOW_DEFAULT_MAX_ROWS)
+
+  def sqlShowDefaultMaxCharsPerTruncatedRow: Int =
```
`sqlShowTruncateMaxCharsPerColumn`?

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:
```diff
@@ -969,6 +969,20 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
   }
 
+  test("SPARK-24442 Show should follow spark.show.default.number.of.rows") {
+    withSQLConf("spark.sql.show.defaultNumRows" -> "100") {
```
`SQLConf.SQL_SHOW_DEFAULT_MAX_ROWS.key -> "100"`

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:
```diff
@@ -969,6 +969,20 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
+  }
+
+  test("SPARK-24442 Show should follow spark.show.default.truncate.characters.per.column") {
+    withSQLConf("spark.sql.show.truncateMaxCharsPerColumn" -> "30") {
```
ditto

maropu commented:
Can you fix the style errors?

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:
```diff
@@ -969,6 +969,20 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
+    withSQLConf("spark.sql.show.truncateMaxCharsPerColumn" -> "30") {
+      (1 to 1000).map(x => "123456789_123456789_123456789_")
+        .toDS().as[String]
+        .show
```
Plz compare the show result?
https://github.com/apache/spark/blob/9241e1e7e66574cfafa68791771959dfc39c9684/sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala#L326

maropu commented:
@AndrewKL Thanks for your work! I think we can add enough tests for this patch without using a spy library.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

mgaido91 commented:
@AndrewKL are you still active on this PR?

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented:
Closing it due to inactivity. Please feel free to take it over.
